content/existential-containers.md

```diff
@@ -0,0 +1,206 @@
+An existential container wraps a value together with a witness of its conformance to one or several type classes into an object exposing the API defined by these type classes.
+
+## Motivation
+
+Type classes can address some of the well-known limitations of subtyping with respect to extensibility, such as the ability to extend existing data types with new behaviors [1].
+A type class describes the interface of a generic _concept_ as a set of requirements, expressed in the form of operations and associated types.
+These requirements can be implemented for a specific type, thereby specifying how this type _models_ the concept.
+The following illustrates:
+
+```scala
+import shapes.{Square, Hexagon}
+
+trait TypeClass:
+  type Self
+
+trait Polygon extends TypeClass:
+  extension (self: Self)
+    def area: Double
+
+given Square is Polygon: ...
+given Hexagon is Polygon: ...
+```
+
+Defining `Polygon` as a type class rather than an abstract class to be inherited allows us to retroactively state that squares are polygons without modifying the definition of `Square`.
+Sticking to subtyping would require the definition of an inefficient and verbose wrapper class.
+
+Alas, type classes offer limited support for eliding type information at compile-time.
+Hence, it is difficult to manipulate heterogeneous collections or write procedures returning arbitrary values known to model a particular concept.
+The following illustrates:
+
+```scala
+def largest[T: Polygon](xs: Seq[T]): Option[T] =
+  xs.maxByOption(_.area)
+
+largest(List(Square(), Hexagon()))
+// error: No given instance of type Polygon{type Self = Square | Hex} was found for a context parameter of method largest
+```
+
+The call to `largest` is illegal because, although there exist witnesses of the `Polygon` and `Hexagon`'s conformance to `Polygon`, no such witness exists for their least common supertype.
+In other words, it is impossible to call `largest` with an heterogeneous sequence of polygons.
+
+## Proposed solution
+
+The problems raised above can be worked around if, instead of using generic parameters with a context bound, we use pairs bundling each value with its conformance witness.
+In broad strokes, our solution generalizes the following possible implementation of `largest`:
+
+```scala
+trait AnyPolygon:
+  type Value: Polygon as witness
+  val value: Value
+
+def largest(xs: Seq[AnyPolygon]): Option[AnyPolygon] =
+  xs.maxByOption((a) => a.witness.area(a.value))
+```
+
+The type `AnyPolygon` conceptually represents an arbitrary polygon.
+It consists of a pair containing some arbitrary value as well as a witness of that value's type being a polygon.
+We call this pair an _existential container_, as a nod to a similar feature in Swift, and the remainder of this SIP explains how to express this idea in a single, type-safe abstraction.
+
+### Specification
+
+Existential containers are encoded as follows:
```

Comment:
It's not clear to me which part of this specification is user-land library code, and which section is compiler/language-level changes. Perhaps we can split it explicitly into two sections?